flang/docs/ParallelFortranRuntime.md

```diff
@@ -0,0 +1,1987 @@
+  in the `sub` argument with the given coarray handle
+    
+```
+subroutine prif_image_index( &
+    coarray_handle, sub, team, team_number, image_index)
+  type(prif_coarray_handle), intent(in) :: coarray_handle
+  integer(c_intmax_t), intent(in) :: sub(:)
+  type(prif_team_type), intent(in), optional :: team
+  integer(c_int), intent(in), optional :: team_number
+  integer(c_int), intent(out) :: image_index
+end subroutine
+```
+**Further argument descriptions**:
+
+* **`team` and `team_number`**: optional arguments that specify a team. They
+  shall not both be present in the same call.
+* **`sub`**: A list of integers that identify a specific image in the
+  identified or current team when interpreted as cosubscripts for the provided
+  coarray handle.
+
+## Coarray Access
+
+The memory consistency semantics of coarray accesses follow those defined
+by the Image Execution Control section of the Fortran standard. In particular,
+coarray accesses will maintain serial dependencies for the issuing image. Any
+data access ordering between images is defined only with respect to ordered
+segments. Note that for put operations, "source completion" means that the provided
+source locations are no longer needed (e.g. their memory can be freed once the procedure
+has returned). 
+
+### Common Arguments
+
+* **`notify_ptr`**: pointer on the identified image to the notify
+  variable that should be updated on completion of the put operation. The
+  referenced variable shall be of type `prif_notify_type`, and the storage
+  must have been allocated using `prif_allocate_coarray` or `prif_allocate`.
+  If this optional argument is omitted, then no notification is performed.
+
+* **`remote_ptr`**: pointer to where on the identified image the data begins.
+  The referenced storage must have been allocated using `prif_allocate_coarray` or `prif_allocate`.
+
+### `prif_put`
+
+**Description**: This procedure assigns to the elements of a coarray, when the elements to be
+  assigned are contiguous in linear memory on both sides. 
+  The compiler can use this to implement assignment to a _coindexed-named-object_. 
+  It need not call this procedure when the coarray reference is not a _coindexed-named-object_. 
+  This procedure blocks on source completion. This procedure corresponds to a contiguous coarray
+  reference on the left hand side of an _assignment-stmt_.
+    
+```
+subroutine prif_put( &
+    coarray_handle, cosubscripts, value, first_element_addr, &
+    team, team_number, notify_ptr, stat, errmsg, errmsg_alloc)
+  type(prif_coarray_handle), intent(in) :: coarray_handle
+  integer(c_intmax_t), intent(in) :: cosubscripts(:)
+  type(*), intent(in), contiguous, target :: value(..)
+  type(c_ptr), intent(in) :: first_element_addr
+  type(prif_team_type), optional, intent(in) :: team
+  integer(c_intmax_t), optional, intent(in) :: team_number
+  integer(c_intptr_t), optional, intent(in) :: notify_ptr
```

Comment:
Quoted from @jeffhammond's response:

> Put with notification is nigh impossible to implement efficiently as specified. Notification variables must be in coarrays. Having the only argument in `prif_put(_raw)` be `notify_ptr` requires an MPI-based implementation (or any other that can't do remote atomic writes to virtual memory directly) to do an expensive lookup on the critical path of what is supposed to be a high-performance operation.
You should implement the put+notify with a separate function that has a second type(`prif_coarray_handle`) argument for the notification variable.

Reply:
**LBL response:** The current specification for `notify_ptr` is motivated by the observation above that the user-specified notification variable access can be either direct or indirect in the corresponding coarray, and a raw-style interface is sufficiently general to capture either scenario. We acknowledge the desire to provide more of the available information (notify coarray handle) to the PRIF implementation, but we’re unsure of the best way to extend the interface to encompass this. In particular, if we take your suggestion as written and double the interface to provide overloads that accept both a `prif_coarray_handle` argument and `notify_ptr` for the notification variable, there is still no guarantee that `notify_ptr` does not reference an indirected component from that coarray, which is a case the implementation would presumably still be obliged to handle.

Reply:
```fortran
program notify_example
   use iso_fortran_env, only: notify_type
   implicit none


   type :: t
       type(notify_type), allocatable :: n
   end type


   type(t) :: notes[*]
   integer :: c[*]


   allocate(notes%n)
   sync all
   if (this_image() == 2) c[1, notify=notes%n] = 42
   if (this_image() == 1) then
       notify wait (notes%n)
       print *, c
   end if
end program
```